core/include/usModuleContext.h

```diff
@@ -1,852 +1,870 @@
+   * specified in its ServiceConstants::SERVICE_ID() property); that is, the
+   * service that was registered first is returned.
+   *
+   * @param clazz The class name with which the service was registered.
+   * @return A <code>ServiceReference</code> object, or an invalid <code>ServiceReference</code> if
+   *         no services are registered which implement the named class.
+   * @throws std::logic_error If this ModuleContext is no longer valid.
+   * @throws ServiceException If no service was registered under the given class name.
+   *
+   * @see #GetServiceReferences(const std::string&, const std::string&)
+   */
+  ServiceReferenceU GetServiceReference(const std::string& clazz);
+
+  /**
+   * Returns a <code>ServiceReference</code> object for a service that
+   * implements and was registered under the specified template class argument.
+   *
+   * <p>
+   * This method is identical to GetServiceReference(const std::string&) except that
+   * the class name for the service object is automatically deduced from the template argument.
+   *
+   * @tparam S The type under which the requested service must have been registered.
+   * @return A <code>ServiceReference</code> object, or an invalid <code>ServiceReference</code> if
+   *         no services are registered which implement the type <code>S</code>.
+   * @throws std::logic_error If this ModuleContext is no longer valid.
+   * @throws ServiceException It no service was registered under the given class name.
+   * @see #GetServiceReference(const std::string&)
+   * @see #GetServiceReferences(const std::string&)
+   */
+  template<class S>
+  ServiceReference<S> GetServiceReference()
+  {
+    std::string clazz = us_service_interface_iid<S>();
+    if (clazz.empty()) throw ServiceException("The service interface class has no US_DECLARE_SERVICE_INTERFACE macro");
+    return ServiceReference<S>(GetServiceReference(clazz));
+  }
+
+  /**
+   * Returns the service object referenced by the specified
+   * <code>ServiceReferenceBase</code> object.
+   * <p>
+   * A module's use of a service is tracked by the module's use count of that
+   * service. Each time a service's service object is returned by
+   * {@link #GetService(const ServiceReference<S>&)} the context module's use count for
+   * that service is incremented by one. Each time the service is released by
+   * {@link #UngetService(const ServiceReferenceBase&)} the context module's use count
+   * for that service is decremented by one.
+   * <p>
+   * When a module's use count for a service drops to zero, the module should
+   * no longer use that service.
+   *
+   * <p>
+   * This method will always return <code>0</code> when the service
+   * associated with this <code>reference</code> has been unregistered.
+   *
+   * <p>
+   * The following steps are taken to get the service object:
+   * <ol>
+   * <li>If the service has been unregistered, <code>0</code> is returned.
+   * <li>The co
```

Comment:
I agree. Based on your and Wayne's feedback, I've revised the entire comment block to

````
  /**
   * Installs a bundle from the specified location.
   *
   * The following steps are required to install a bundle:
   * -# If a bundle containing the same location identifier is already installed, the Bundle object for that 
   *    bundle is returned.
   * -# The bundle's associated resources are allocated. The associated resources minimally consist of a
   *    unique identifier and a persistent storage area if the platform has file system support. If this step
   *    fails, a std::runtime_error is thrown.
   * -# A bundle event of type <code>BundleEvent::INSTALLED</code> is fired.
   * -# The Bundle object for the newly or previously installed bundle is returned.
   *
   * @remarks A location identifier is defined as an absolute path to a shared library or executable file
   * followed by a slash (/) and the bundle's name.
   * 
   * For example: 
   * -# <code>InstallBundle("/path/to/bundle/foo.so/foo");</code>
   * -# <code>InstallBundle("/path/to/bundle/foo.so/my_static_bundle");</code>
   *
   * @param location The location identifier of the bundle to install.
   * @return The Bundle object of the installed bundle.
   * @throws std::runtime_error If the installation failed.
   */
````